src/Tools/DynamoShapeManager/Utilities.cs

```diff
@@ -202,6 +202,58 @@ public static Version GetInstalledAsmVersion2(IEnumerable<Version> versions, ref
             return null;
         }
 
+        /// <summary>
+        /// Get the corresponding libG preloader location for the target ASM loading version.
+        /// If there is exact match preloader version to the target ASM version, use it, 
+        /// otherwise use the closest below.
+        /// </summary>
+        /// <param name="version">The target loading version of ASM.</param>
+        /// <param name="rootFolder">Full path of the directory that contains 
+        /// LibG_major_minor_build folder. In a 
+        /// typical setup this would be the same directory that contains Dynamo 
+        /// core modules. This must represent a valid directory.</param>
+        /// <returns></returns>
+        public static string GetLibGPreloaderLocation(Version version, string rootFolder)
+        {
+            if (version == null)
+            {
+                version = new Version(0, 0, 0);
+            }
+
+            var libGFolderName = string.Format("libg_{0}_{1}_{2}", version.Major, version.Minor, version.Build);
+            var dir = new DirectoryInfo(Path.Combine(rootFolder, libGFolderName));
+            if (dir.Exists)
+            {
+                return dir.FullName;
+            }
+            else
+            {
+                // This usually means libG preloader version is behind the target version
+                var rootDir = new DirectoryInfo(rootFolder);
+
+                // Use regex to get all the libG versions supported
+                var libgFolders = rootDir.EnumerateDirectories("libg_*", SearchOption.TopDirectoryOnly);
+                var regExp = new Regex(@"^libg_(\d\d\d)_(\d)_(\d)$", RegexOptions.IgnoreCase);
+                var preloaderVersions = new List<Version>();
+                foreach (var folder in libgFolders)
+                {
+                    var match = regExp.Match(folder.Name);
+                    if (match.Groups.Count == 4 && Convert.ToInt32(match.Groups[1].Value) <= version.Major)
+                    {
+                        preloaderVersions.Add(new Version(
+                                Convert.ToInt32(match.Groups[1].Value),
+                                Convert.ToInt32(match.Groups[2].Value),
+                                Convert.ToInt32(match.Groups[3].Value)));
+                    }
+                }
+                preloaderVersions.Sort();
+                preloaderVersions.Reverse();
+                // Pick the closest preloader version below or use the default value when no libG folder found
+                var preloaderVersion = preloaderVersions.FirstOrDefault() == null ? version : preloaderVersions.First();
+                return Path.Combine(rootFolder, string.Format("libg_{0}_{1}_{2}", preloaderVersion.Major, preloaderVersion.Minor, preloaderVersion.Build));
```

Comment:
If there's no matching libg folder found for the target `version` passed in, then setting `preloaderVersion` to `version` and constructing a libg folder path from it will return a non-existent location.

Reply:
@aparajit-pratap `Preload()` will fail in this case, do you prefer just fall back to whatever libG folder available even the version is larger?

Reply:
@aparajit-pratap If we end up returning an invalid LibG folder although unlikely, we will throw exception at https://github.com/DynamoDS/Dynamo/blob/7b885b76ea585128af0eb997e0f1567545952466/src/Tools/DynamoShapeManager/Utilities.cs#L289 which will end up in the crash dialog and analytics. I can make the exception message more explicit